File: transpiler/javatests/com/google/j2cl/integration/java/superfieldaccess/Main.java
/*
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package superfieldaccess;

import static com.google.j2cl.integration.testing.Asserts.assertEquals;

/** Test field access using the {@code super} keyword. */
public class Main {
  public static void main(String[] args) {
    testSuperFieldAccess();
  }

  private static void testSuperFieldAccess() {
    class GrandParent {
      public String name = "GrandParent";
    }

    class Parent extends GrandParent {
      final String name;

      Parent(String from) {
        this.name = "Parent (from " + from + ")";
      }

      public String getParentName() {
        return super.name;
      }
    }

    class Child extends Parent {
      final String name;

      Child() {
        this("Child");
      }

      Child(String from) {
        super(from);
        this.name = "Child (from " + from + ")";
      }

      public String getParentName() {
        return super.name;
      }

      public String getGrandParentName() {
        return super.getParentName();
      }

      class Inner extends Parent {
        Inner() {
          super("Inner");
        }

        public String getOuterParentName() {
          return Child.super.name;
        }

        class InnerInner {
          public String getOuterParentNameQualified() {
            return Child.Inner.super.name;
          }
        }
      }

    }

    assertEquals("Parent (from Child)", new Child().getParentName());
    assertEquals("GrandParent", new Child().getGrandParentName());
    assertEquals("Parent (from Child)", new Child().new Inner().getOuterParentName());
    assertEquals(
        "Parent (from Inner)",
        new Child().new Inner().new InnerInner().getOuterParentNameQualified());
  }
}

File: transpiler/javatests/com/google/j2cl/integration/kotlin/superfieldaccess/Main.kt
/*
 * Copyright 2022 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package superfieldaccess

import com.google.j2cl.integration.testing.Asserts.assertEquals

/** Test field access using the `super` keyword. */
fun main(vararg unused: String) {
  testSuperFieldAccess()
}

private fun testSuperFieldAccess() {
  open class GrandParent {
    open var name = "GrandParent"
  }

  open class Parent(from: String) : GrandParent() {
    override var name = "Parent (from " + from + ")"

    open fun getParentName(): String {
      return super.name
    }
  }

  class Child(from: String) : Parent(from) {
    constructor() : this("Child")

    override var name = "Child (from " + from + ")"

    override fun getParentName(): String {
      return super.name
    }

    fun getGrandParentName(): String {
      return super.getParentName()
    }

    inner class Inner : Parent("Inner") {
      fun getOuterParentName(): String {
        return super@Child.name
      }

      inner class InnerInner {
        fun getOuterParentNameQualified(): String {
          // In here to keep in sync with the Java version. Kotlin does not allow to use
          // a qualified name to qualify a super reference.
          return super@Inner.name
        }
      }
    }
  }
  assertEquals("Parent (from Child)", Child().getParentName())
  assertEquals("GrandParent", Child().getGrandParentName())
  assertEquals("Parent (from Child)", Child().Inner().getOuterParentName())
  assertEquals("Parent (from Inner)", Child().Inner().InnerInner().getOuterParentNameQualified())
}
